Feature/yyjd 165 action card api impl

jooho172200 opened the pull request:
작업 내용
- 카드 도메인 구조 수정
- 도구/길 카드 서비스 클래스 구현
- 카드 사용 이벤트 등록 클래스 구현
- 카드 dto 클래스 구현
- 테스트용 카드 초기화 코드 작성

jooho172200 committed:
feat: 카드 서비스 및 이벤트 처리 로직 구현

- 도구/길 카드 서비스 클래스 구현
- 카드 타입 관련 enum(CardType) 추가
- 카드 도메인에 타입 정보 및 기능 메서드 추가
- 카드 사용 DTO(UseCardRequest, UseCardResponse) 생성
- 카드 이벤트 처리용 CardEventRegister 작성

jooho172200 committed:
merge: develop 병합

jooho172200 committed:
- ActionCardService 테스트용 코드 추가
- 일부 클래스에 테스트를 위해 @Setter 추가

zido0938 commented:
어떻게 테스트를 진행하신지 설명이 좀 부족하긴한데 NULL값이 나오는 이유에 대해 제가 추측해서 말씀을 드리자면, 현재 주호님 코드에서
public Card() {
        //this.cardId = UUID.randomUUID();
    } 해당부분을 주석처리하신 이유가 아마 각 카드마다 UUID를 직접 지정해서 쓰시려고 하신 것 같은데, 이러면 GameCardPool을 보시면 createDefaultPool 함수에서 길카드, 행동카드를 default 카드덱에 추가할때 UUID가 생성되지않고 Null값이 들어갈 겁니다.(기존 제 코드에서는 Card 클래스에서 UUID를 랜덤으로 생성하기때문에 해당 함수 동작에서 각 카드마다 랜덤 UUID가 생성됨)
따라서 만약 카드에 지정된 UUID를 넣어서 postman에서 동작을 테스트하고 싶으시다면 테스트코드에서 setCard 하신 것 처럼 카드풀을 생성 시 UUID를 주입해줘야 하지 않을까 생각됩니다. 

추가로 스크린샷만보면 웹소켓 연결 후 바로 카드 사용을 테스트하시는 것 처럼 보이는데, 게임유저와 방을 생성하고 게임을 시작하면 유저들에게 카드가 나눠질텐데 해당 과정없이 바로 카드 사용을 테스트하는 것이 맞는지는 잘 모르겠습니다.

jooho172200 commented:
> 어떻게 테스트를 진행하신지 설명이 좀 부족하긴한데 NULL값이 나오는 이유에 대해 제가 추측해서 말씀을 드리자면, 현재 주호님 코드에서 public Card() { //this.cardId = UUID.randomUUID(); } 해당부분을 주석처리하신 이유가 아마 각 카드마다 UUID를 직접 지정해서 쓰시려고 하신 것 같은데, 이러면 GameCardPool을 보시면 createDefaultPool 함수에서 길카드, 행동카드를 default 카드덱에 추가할때 UUID가 생성되지않고 Null값이 들어갈 겁니다.(기존 제 코드에서는 Card 클래스에서 UUID를 랜덤으로 생성하기때문에 해당 함수 동작에서 각 카드마다 랜덤 UUID가 생성됨) 따라서 만약 카드에 지정된 UUID를 넣어서 postman에서 동작을 테스트하고 싶으시다면 테스트코드에서 setCard 하신 것 처럼 카드풀을 생성 시 UUID를 주입해줘야 하지 않을까 생각됩니다.
> 
> 추가로 스크린샷만보면 웹소켓 연결 후 바로 카드 사용을 테스트하시는 것 처럼 보이는데, 게임유저와 방을 생성하고 게임을 시작하면 유저들에게 카드가 나눠질텐데 해당 과정없이 바로 카드 사용을 테스트하는 것이 맞는지는 잘 모르겠습니다.

아 저는 웹소켓 테스트용으로 작성한 ActionCardServiceTest를 사용했습니다. json 요청을 보낼 때 id를 지정해서 보내기 위해 잠시 UUID를 랜덤으로 발급해주는 부분을 주석 처리한 거였습니다.

jooho172200 committed:
- ActionCardServiceTest 작성

jooho172200 committed:
- merge: merged develop

jooho172200 committed:
refactor: 최신 GameCardPool, UserCardDeck 구조에 맞춰 카드 서비스 로직 수정

jooho172200 commented:
최신 카드풀과 유저덱 클래스들을 사용해서 테스트했더니 이제 카드 ID를 찾을 수 없다는 오류가 나지 않습니다. 나머지 웹소켓 테스트들 진행 후 최대한 빠르게 올려두겠습니다.

jooho172200 committed:
refactor: 에러 코드 수정

jooho172200 commented:
기존 웹소켓 테스트용 클래스 명을 ActionCardServiceTest.java -> ActionCardServiceInitTest.java로 변경하였습니다

jooho172200 commented:
제가 그동안 착각을 했었네요 웹소켓 테스트 정상적으로 되는 것 같습니다. 나머지 부분도 테스트해서 올리겠습니다

jooho172200 commented:
제가 웹소켓 테스트를 잘못하고 있었던 거 같습니다. 도구 카드 사용 확인 되었고 나머지도 테스트 후 푸시하겠습니다.
![image](https://github.com/user-attachments/assets/d086826a-71a9-4228-8617-30c6fe9ec836)

jooho172200 committed:
feat: useCard ack 응답 처리 및 길카드 테스트용 WebSocket 로직 추가
style: spotless 적용

jooho172200 committed:
feat: 카드 요청 DTO 분리 및 서비스 구조 개선

- 유저 대상 Action 카드(BREAK, REPAIR)용 DTO와 셀 대상 카드(MAP, FALLING_ROCK)용 DTO 분리
- 길카드 전용 PathCardRequest DTO 추가
- UserCardDeck에 카드 ID 기반 카드 조회 메서드(getCardById) 추가
- ActionCardService 내 카드 타입 분기에 따라 서비스 호출 구조 변경
- WebSocket 이벤트에서 카드 타입별 useCard 이벤트 명 분리 (ex. useBreakToolCard 등)
